src/code/cutscene_camera.c

```diff
@@ -353,28 +357,579 @@ void CutsceneCamera_Reset(void) {
     sCurCsCamera->state = CS_CAM_STATE_UPDATE_ALL;
 }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/cutscene_camera/func_80161C20.s")
+// Linear interpolation from initial values to cmd values. Set weight to 100 to go all the way to the cmd value.
+s16 CutsceneCamera_Interp_Lerp(Vec3f* camPos, f32* camFov, s16* camRoll, CsCmdCamPoint* pointCmd, CsCmdCamMisc* miscCmd,
+                               CutsceneCameraInterp* interpState) {
+    f32 lerp;
+
+    if (interpState->type != CS_CAM_INTERP_LERP) {
+        // Initialize
+        interpState->type = CS_CAM_INTERP_LERP;
+        interpState->waypoint = 0;
+        interpState->curFrame = 0;
+        interpState->duration = 1;
+        if (camPos != NULL) {
+            interpState->initPos.x = camPos->x;
+            interpState->initPos.y = camPos->y;
+            interpState->initPos.z = camPos->z;
+        }
+
+        if (camFov != NULL) {
+            interpState->initFov = *camFov;
+        }
+
+        if (camRoll != NULL) {
+            interpState->initRoll = *camRoll;
+        }
+    }
+
+    interpState->curFrame++;
+
+    lerp = ((f32)interpState->curFrame / pointCmd->duration) * (pointCmd->weight / 100.0f);
+
+    if (camPos != NULL) {
+        VEC3F_LERPIMPDST(camPos, &interpState->initPos, &pointCmd->pos, lerp);
+    }
+
+    if (camFov != NULL) {
+        *camFov = LERPIMP(interpState->initFov, miscCmd->fov, lerp);
+    }
+
+    if (camRoll != NULL) {
+        s16 new_var;
+        s16 temp;
+
+        new_var = CAM_DEG_TO_BINANG(miscCmd->roll);
+
+        temp = (s16)(new_var - (s16)interpState->initRoll);
+
+        *camRoll = (s16)interpState->initRoll + (s16)(temp * lerp);
+
+        // *camRoll = LERPIMP((s16)interpState->initRoll, (s16)CAM_DEG_TO_BINANG(miscCmd->roll), lerp);
+    }
+
+    if (interpState->curFrame >= pointCmd->duration) {
+        // Finished
+        interpState->type = CS_CAM_INTERP_OFF;
+        return 1;
+    }
+
+    return 0;
+}
+
+s16 CutsceneCamera_Interp_Smooth(Vec3f* camPos, f32* camFov, s16* camRoll, CsCmdCamPoint* pointCmd,
+                                 CsCmdCamMisc* miscCmd, CutsceneCameraInterp* interpState) {
+    f32 lerp;
+    f32 tmp1;
+    f32 tmp2;
+
+    if (interpState->type != CS_CAM_INTERP_SMOOTH) {
+        // Initialize
+        interpState->type = CS_CAM_INTERP_SMOOTH;
+        interpState->waypoint = 0;
+        interpState->curFrame = 0;
+        interpState->duration = 1;
+        if (camPos != NULL) {
+            interpState->initPos.x = camPos->x;
+            interpState->initPos.y = camPos->y;
+            interpState->initPos.z = camPos->z;
+        }
+        if (camFov != NULL) {
+            interpState->initFov = *camFov;
+        }
+        if (camRoll != NULL) {
+            interpState->initRoll = *camRoll;
+        }
+    }
+
+    // could rewrite `(pointCmd->duration & 1)` as `(s32)((u32)pointCmd->duration % 2)`
```

Comment:
Which would you prefer? I'm not 100% sure this comment is necessary; plenty of functions have multiple ways of matching code, and we don't always write these different ways down in a comment like this. We just pick the one we like and stick with it.